core/src/main/scala/cats/data/AndThen.scala

```diff
@@ -0,0 +1,112 @@
+package cats.data
+
+import java.io.Serializable
+
+/**
+ * A function type of a single input that can do function composition
+ * (via `andThen` and `compose`) in constant stack space with amortized
+ * linear time application (in the number of constituent functions).
+ *
+ * Example:
+ *
+ * {{{
+ *   val seed = AndThen((x: Int) => x + 1))
+ *   val f = (0 until 10000).foldLeft(seed)((acc, _) => acc.andThen(_ + 1))
+ *
+ *   // This should not trigger stack overflow ;-)
+ *   f(0)
+ * }}}
+ */
+private[cats] sealed abstract class AndThen[-T, +R]
+  extends (T => R) with Product with Serializable {
+
+  import AndThen._
+
+  final def apply(a: T): R =
+    runLoop(a)
+
+  override def andThen[A](g: R => A): AndThen[T, A] = {
+    // Fusing calls up to a certain threshold, using the fusion
+    // technique implemented for `cats.effect.IO#map`
+    this match {
+      case Single(f, index) if index != 127 =>
+        Single(f.andThen(g), index + 1)
+      case _ =>
+        andThenF(AndThen(g))
+    }
+  }
+
+  override def compose[A](g: A => T): AndThen[A, R] = {
+    // Fusing calls up to a certain threshold, using the fusion
+    // technique implemented for `cats.effect.IO#map`
+    this match {
+      case Single(f, index) if index != 127 =>
```

Comment:
Nitpick, can we make this 127 (and at line 32) a constant with a sensible name?

Reply:
Thanks for the suggestion. I've updated the code with a documented constant.